File: lua/day10.lua
local utility = require("utility")

local Day10 = {
    path = "../input/day10.txt"
}

function Day10:run()
    local f = assert(io.open(self.path, "r"))
    local s = f:read("*all")
    f:close()
    local instructions = self.make_instructions(s)
    local part1, part2 = self.solve(instructions)
    print("Part 1:\n" .. part1 .. "\n")
    print("Part 2:\n" .. part2 .. "\n")
end

function Day10.make_instructions(s)
    local instructions = {}

    for line in s:gmatch("([^\n]*)\n?") do
        local i, j = string.find(line, "-?%d+")

        if i then
            local val = line:sub(i, j)
            table.insert(instructions, {cycles = 2, value = tonumber(val)})
        else
            table.insert(instructions, {cycles = 1, value = 0})
        end
    end
    return instructions
end

function Day10.solve(ins)
    local cycle, pixels, pos, x, sum = 1, "", 0, 1, 0

    while #ins > 0 do
        if (cycle - 20) % 40 == 0 then sum = sum + x * cycle end

        -- Draw the pixel
        local pixel = "."
        if pos >= x - 1 and pos <= x + 1 then pixel = "#" end
        pixels, pos = pixels .. pixel, pos + 1

        -- Reset position if this is end of the CRT's draw distance
        if pos % 40 == 0 then
            pixels, pos = pixels .. "\n", 0
        end

        -- Apply instruction to move the register
        ins[1].cycles = ins[1].cycles - 1
        if ins[1].cycles <= 0 then
            x = x + ins[1].value
            table.remove(ins, 1)
        end
        cycle = cycle + 1
    end
    return sum, pixels
end

return Day10

File: 2022/swift/Sources/AdventOfCode/Subcommands/Day10.swift
import ArgumentParser
import Foundation

struct Day10: ParsableCommand {
    static let configuration = CommandConfiguration(abstract: "Advent of Code - 2022 December 10", version: "1.0.0")

    // MARK - Data structures

    struct CycleInstruction {
        var cycles: Int
        let value: Int

        init?(_ rawValue: String) {
            let components = rawValue.components(separatedBy: .whitespaces)
            switch components[0].lowercased() {
            case "noop":
                self.cycles = 1
                self.value = 0
            case "addx":
                self.cycles = 2
                self.value = components.count == 2 ? Int(components[1]) ?? 0 : 0
            default:
                return nil
            }
        }
    }

    // MARK: - Options

    @Option(name: .shortAndLong, help: "Input file path")
    var path: String = "../input/day10.txt"

    // MARK: - Lifecycle

    mutating func run() throws {
        let instructions = makeInstructions(fromString: try String(contentsOfFile: path))
        let (part1, part2) = solve(instructions: instructions)
        print("Part 1:\n\(part1)\n")
        print("Part 2:\n\(part2)")
    }

    // MARK: - Internal methods

    func makeInstructions(fromString str: String) -> [CycleInstruction] {
        str.components(separatedBy: .newlines).compactMap(CycleInstruction.init)
    }

    func solve(instructions: [CycleInstruction]) -> (Int, String) {
        var ins = instructions
        var (cycle, pixels, pos, x, sum) = (1, "", 0, 1, 0)

        while ins.count > 0 {
            if (cycle - 20) % 40 == 0 {
                sum += x * cycle
            }

            // Draw the pixel
            pixels += x-1...x+1 ~= pos ? "#" : "."
            pos += 1

            // Reset position if this is end of the CRT's draw distance
            if pos % 40 == 0 {
                pixels += "\n"
                pos = 0
            }

            // Apply instructions to move the register
            ins[0].cycles -= 1

            if ins[0].cycles <= 0 {
                x += ins[0].value
                ins.removeFirst()
            }
            cycle += 1
        }
        return (sum, pixels)
    }
}